torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -886,68 +929,122 @@ void ProcessGroupNCCL::logWorkEnd(WorkNCCL& work) {
+                  << " Please file an issue.";
+            }
           }
+          // Throw exception
+          work.handleException(asyncErrorHandling_);
         }
-        // Throw exception
-        work.handleException(asyncErrorHandling_);
-      }
 
-      // Work status logging for desync debug
-      if (desyncDebug_) {
-        if (work.isStarted()) {
-          logWorkStart(work);
+        // Work status logging for desync debug
+        if (desyncDebug_) {
+          if (work.isStarted()) {
+            logWorkStart(work);
+          }
+          if (work.isCompleted()) {
+            logWorkEnd(work);
+          }
         }
+
+        // Clean up completed work
         if (work.isCompleted()) {
-          logWorkEnd(work);
+          if (onCompletionHook_) {
+            // onCompletion hook will consume the work, so keep them alive in a
+            // separate list for now
+            completedWorkList.splice(
+                completedWorkList.begin(), workMetaList_, it++);
+          } else {
+            it = workMetaList_.erase(it);
+          }
+        } else {
+          // Increment the iterator if the current WorkNCCL object is not
+          // completed.
+          ++it;
         }
       }
+      done = workMetaList_.empty();
+    }
+    // Python hook will grab GIL. Release workMetaListMutex_ before acquiring
+    // GIL to avoid potential deadlock.
 
-      // Clean up completed work
-      if (work.isCompleted()) {
-        it = workMetaList_.erase(it);
-      } else {
-        // Increment the iterator if the current WorkNCCL object is not
-        // completed.
-        ++it;
+    try {
+      while (!completedWorkList.empty()) {
+        // The work object will be passed to Python side. Wrap it
+        // with an intrusive_ptr so that pybind can safely pass by copy
+        // and own the ptr lifetime.
+        onCompletionHook_(
+            c10::make_intrusive<WorkNCCL>(completedWorkList.back()));
+        completedWorkList.pop_back();
+      }
+    } catch (std::exception& e) {
+      // PythonOnCompletionHook has already extracted Python exception message
+      // and wrapped it with a cpp one. So we no longer need to acquire GIL
+      // here.
+      const auto errorStr = c10::str(
+          "Caught exception while running onCompletion hook for ProcessGroupNCCL: ",
+          e.what(),
+          "Aborting all communicators.");
+
+      if (!completedWorkList.empty()) {
+        // CompletedWorkList is not empty, so it should be the first work in
+        // the queue triggered this exception. Use that work object to handle
+        // the error.
+        completedWorkList.back().setException(
+            std::make_exception_ptr(std::runtime_error(errorStr)));
+        completedWorkList.back().abort();
+        completedWorkList.back().handleException(asyncErrorHandling_);
       }
+      // Abort all NCCL Communicators on this ProcessGroupNCCL instance.
+      abort();
```

Comment:
in what case would `completedWorkList` be empty / do we ever get to this `abort()`?

Reply:
It should never get here, but feels un-secure if we just ignore the else branch. Let me add an `abortReason` argument to clarify in case somehow this branch is hit.

Reply:
addressed